api.go

```diff
@@ -830,3 +834,112 @@ func (c Client) makeTargetURL(bucketName, objectName, bucketLocation string, que
 
 	return u, nil
 }
+
+// isVirtualHostStyle - returns true if server supports
+// virtual hosted-style urls
+func (c *Client) isVirtualHostStyle(bucketName string) bool {
+	if c.endpointURL == sentinelURL {
+		return false
+	}
+	// bucketName can be valid but '.' in the hostname will fail SSL
+	// certificate validation. So do not use host-style for such buckets.
+	if c.endpointURL.Scheme == "https" && strings.Contains(bucketName, ".") {
+		return false
+	}
+	return c.isVirtual
+}
+
+// probeVirtualHostStyle - sets the isVirtualHostStyle bool
+// to true if server can accept virtual hosted style urls.
+// Send a HEAD bucket request for a random bucket name
+// with request url in virtual hosted style format,
+// and see if server responds with NotFound status.
+func (c *Client) probeVirtualHostStyle() {
+	// set default to false
+	c.isVirtual = false
+
+	host := c.endpointURL.Host
+	scheme := c.endpointURL.Scheme
+	method := "HEAD"
+	// Strip port 80 and 443 so we won't send these ports in Host header.
+	if h, p, err := net.SplitHostPort(host); err == nil {
+		if scheme == "http" && p == "80" || scheme == "https" && p == "443" {
+			host = h
+		}
+	}
+	// Generate a new random bucket name.
```

Comment:
### Motivation
How about we simplify the heuristic to determine if virtual host style access is supported by the object storage service? Apologies for making this recommendation this late in the review.
My aim is,
1. to ensure that we *never* incorrectly detect virtual host support but can incorrectly fall back to path style access. 
2. to avoid deciding which signature version to use in `privateNew`

### Suggested approach
One way would be to simply make an anonymous [GET bucket location](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGETlocation.html) request to a constant bucket name. There are 3 outcomes which indicate virtual host style support.

1. Bucket Found 200 OK (valid location XML response) - Bucket owned by user
(Only possible if there is a [bucket policy](https://docs.aws.amazon.com/AmazonS3/latest/dev/using-with-s3-actions.html#using-with-s3-actions-related-to-buckets) allowing GetBucketLocation operation on the bucket)
2. Bucket Not Found XML error response
3. Access Denied XML error response - Bucket owned by someone else  

For all other responses, we can safely fall back to path-style.

### Remarks
N B in the current approach, using HEAD bucket request we can't differentiate between an S3 bucket and a plain HTTP asset hosted at `bucketname.endpoint`, from a 200 OK response. This is why I recommend GET bucket location, which returns an XML response on 200 OK as well.

Reply:
@krisis, how would we guarantee a constant bucket name , short of creating one while probing virtual hosted style support? The other option going by your approach would be to listbuckets and then pick the first bucket and do a HEAD bucket or GET location request - that way, the bucket is guaranteed to belong to user and we would revert to path style only if the server has no buckets created so far / bucket got deleted before HEAD operation is attempted. Thoughts?

Reply:
> how would we guarantee a constant bucket name , short of creating one while probing virtual hosted style support? 

We need __not__ create a bucket with a fixed name. We can use any arbitrarily chosen (fixed) bucket name and perform an anonymous GET bucket location request. The heuristic in my previous comment describes how we handle all possible responses.

Reply:
@krisis, user may not have GET Bucket Location permission - in which case this has same pitfalls as   the HEAD bucket probing.

Reply:
@poornas, for the S3 like service to determine that GET bucket location or HEAD request is not allowed requires the service to resolve the bucket from the host or request resource. In virtual host style access, it is determined from subdomain name of hostname. So, we can infer that the service should have determined the bucket from hostname.

Reply:
@krisis, tested your heuristic locally - the AccessDenied error can be raised when the bucket is owned by someone else as well as if the server doesn't support virtual host style (for e.g. while running against localhost). On the other hand, if we excluded AccessDenied in the heuristic we would miss out cases where the server  supports virtual hosted style but the bucket policy doesnt allow anonymous access.